### ISSUE TITLE
View MapFunction incorrect date parsing

### ISSUE THREAD

**Issue by tudordumitriu** [apache/couchdb#1295]
Inside a Map Function when trying to parse a date an extra day is added.
## Expected Behavior

2018-04-26T14:00:00.000Z | x  
2018-04-26T14:00:00.000Z | y

## Current Behavior

2018-04-26T14:00:00.000Z | x  
2018-04-27T14:00:00.000Z | y

## Steps to Reproduce (for bugs)

`   emit(doc.startDateTime, "x");
   var currentDateTime = new Date(doc.startDateTime);
   emit(currentDateTime, "y");`
1.
2.
3.
4.

## Your Environment

* Version used: 2.1.1
* Browser Name and version:
* Operating System and version (desktop or mobile): Windows 10
* Link to your project:

**Comment by tudordumitriu** [comment]
Hi
I have attached the full map function (a simplified version but nevertheless buggy)
```
function (doc) {
         if(doc.startDateTime){ 
             var currentDateTime = new Date(doc.startDateTime);          
             emit(currentDateTime, 0);
             var currentOccurenceNo = 0;
             while (currentOccurenceNo < 4) 
             {
                currentDateTime = new Date(currentDateTime.setDate(currentDateTime.getDate() + 1));
                emit(currentDateTime, 0);
                currentOccurenceNo ++;
             }
         }
}
```
For the document is enough to have a document with startDateTime: "2018-04-24T14:00:00.000Z" and the results are the following:


9d18eb2eaf35aed06b5ef334d702e39c | 2018-04-25T14:00:00.000Z | 0 |  
9d18eb2eaf35aed06b5ef334d702e39c | 2018-04-26T14:00:00.000Z | 0 |  
9d18eb2eaf35aed06b5ef334d702e39c | 2018-04-27T14:00:00.000Z | 0 |  
9d18eb2eaf35aed06b5ef334d702e39c | 2018-04-28T14:00:00.000Z | 0 |  
9d18eb2eaf35aed06b5ef334d702e39c | 2018-04-28T14:00:00.000Z | 0 |

Also, VERY IMPORTANT I THINK, if we comment out the while loop, the initial emit is displaying correctly 
9d18eb2eaf35aed06b5ef334d702e39c | 2018-04-24T14:00:00.000Z | 0 |

Hope this will help more

**Comment by eiri** [comment]
Yes, as @willholley said the issue here is that you are mutating `currentDateTime` within your function. The subtlety here is that `emit` doesn't return immediately from a map function, but rather pushes the emitted pairs on an accumulator that processed after. Since objects are a reference type in javascript changing the original values changes them on acc too. You need either allocate new Date object or stringify the keys before emitting.

**Comment by tudordumitriu** [comment]
Thanks guys!
